casper/src/main/scala/coop/rchain/casper/Validate.scala

```diff
@@ -1,40 +1,96 @@
 package coop.rchain.casper
 
-import cats.Monad
+import cats.Applicative
 import cats.implicits._
 
 import com.google.protobuf.ByteString
 
-import coop.rchain.casper.protocol.BlockMessage
+import coop.rchain.casper.Estimator.{BlockHash, Validator}
+import coop.rchain.casper.protocol.{BlockMessage, Justification}
 import coop.rchain.casper.util.ProtoUtil
 
 import coop.rchain.crypto.hash.Blake2b256
 import coop.rchain.crypto.signatures.Ed25519
 
 import coop.rchain.p2p.effects.Log
 
+import scala.util.Try
+
 object Validate {
   private def ignore(b: BlockMessage, reason: String): String =
     s"CASPER: Ignoring block ${PrettyPrinter.buildString(b.blockHash)} because $reason"
 
-  def blockSignature[F[_]: Monad: Log](block: BlockMessage): F[Boolean] =
-    block
-      .pure[F]
-      .flatMap(b => {
-        if (b.sigAlgorithm == "ed25519") {
-          val justificationHash = ProtoUtil.protoSeqHash(b.justifications)
-          val sigData           = Blake2b256.hash(justificationHash.toByteArray ++ b.blockHash.toByteArray)
-          val isValid           = Ed25519.verify(sigData, b.sig.toByteArray, b.sender.toByteArray)
-
-          if (isValid) {
-            true.pure[F]
-          } else {
-            Log[F].warn(ignore(b, "signature is invalid.")) *> false.pure[F]
-          }
-        } else {
-          Log[F].warn(
-            ignore(b, s"signature algorithm ${b.sigAlgorithm} is unsupported.")
-          ) *> false.pure[F]
-        }
+  def blockSignature[F[_]: Applicative: Log](b: BlockMessage): F[Boolean] =
+    if (b.sigAlgorithm == "ed25519") {
+      val justificationHash = ProtoUtil.protoSeqHash(b.justifications)
+      val sigData =
+        Blake2b256.hash(justificationHash.toByteArray ++ b.blockHash.toByteArray)
+      val isValid =
+        Try(Ed25519.verify(sigData, b.sig.toByteArray, b.sender.toByteArray)).getOrElse(false)
+
+      if (isValid) {
+        true.pure[F]
+      } else {
+        Log[F].warn(ignore(b, "signature is invalid.")) *> false.pure[F]
+      }
+    } else {
+      Log[F].warn(
+        ignore(b, s"signature algorithm ${b.sigAlgorithm} is unsupported.")
+      ) *> false.pure[F]
+    }
```

Comment:
Hmm so if this is safe, what was the issue that made us cautious about wrapping everything in Capture?

Reply:
Pawel said he'll teach us tomorrow unless you already have a solid understanding.

Reply:
The point of `F` is to capture effects. So you do not need to wrap something in a `F` when it is a pure function. Here, I do some pure checks against the input values and the only effect is the logging, which is why we don't need any `F` until the end when we write the log message and return the result. What I had done wrong before was pretty niche and I over reacted by wrapping everything in `F`s. The problem there was that I was producing a "recipe" (as Pawel says) and then calling `.forever` on it, but an important check was not contained in the recipe, so the check was not repeated.

Reply:
And thanks for the explanation